drivers/volume/portworx/portworx.go

```diff
@@ -1684,6 +1722,94 @@ func (d *portworx) CollectDiags(n node.Node) error {
+		if err != nil {
+			return nil, true, err
+		}
+		nodePools := d.getPoolTotalSizes(cluster.Nodes)
+		for _, nodePool := range *nodePools {
+			if err = d.validatePoolSize(nodePool); err != nil {
+				return nil, true, err
+			}
+		}
+		return nil, false, nil
+	}
+	_, err := task.DoRetryWithTimeout(t, timeout, retryInterval)
+	return err
+}
+
+func (d *portworx) getPoolTotalSizes(nodes []api.Node) *map[string]storagepool {
+	nodePools := make(map[string]storagepool)
+	for _, n := range nodes {
+		nodePools[n.SchedulerNodeName] = storagepool{
+			NodeName:  n.SchedulerNodeName,
+			PoolSizes: d.getPools(n),
+		}
+	}
+	return &nodePools
+}
+
+func (d *portworx) validatePoolSize(nodePool storagepool) error {
+	initialNodeState, ok := d.initialClusterSize[nodePool.NodeName]
+	if !ok {
+		return fmt.Errorf("initial node state not found for: %s", nodePool.NodeName)
+	}
+	for i, currentPoolSize := range nodePool.PoolSizes {
+		if d.PoolScalePercentageUsage < d.PoolScaleMaxSize*100/int64(currentPoolSize) {
+			calculatedPoolSize := int64(float64(initialNodeState.PoolSizes[i]) * (float64(d.PoolScalePercentageUsage)/100 + 1))
+			if int64(currentPoolSize) < calculatedPoolSize {
+				return fmt.Errorf("pool size differs. Expected: %d Got: %d", calculatedPoolSize, currentPoolSize)
+			}
+		}
+	}
+	return nil
+}
+
+func (d *portworx) createAutopilotPoolExpandObject(apParams *torpedovolume.AutopilotRuleParams) (*ap_api.AutopilotRule, error) {
+	obj := &ap_api.AutopilotRule{}
+	obj.Name = apParams.Name
+	obj.Spec.Enforcement = ap_api.EnforcementRequired
+
+	actions := &ap_api.RuleAction{
+		Name: "penstorage.io.action.storagepool/expand",
+		Params: map[string]string{
+			"scalepercentage": strconv.FormatInt(apParams.PoolScalePercentageUsage, 10),
+			"scaletype":       "resize-disk",
+		},
+	}
+	obj.Spec.Actions = append(obj.Spec.Actions, actions)
+
+	expPoolAvailableCapacity := &ap_api.LabelSelectorRequirement{
+		Key:      "100 * ( px_pool_stats_available_bytes/ px_pool_stats_total_bytes)",
+		Operator: "Lt",
+		Values:   []string{strconv.FormatInt(apParams.PoolScalePercentageUsage, 10)},
+	}
+	obj.Spec.Conditions.Expressions = append(obj.Spec.Conditions.Expressions, expPoolAvailableCapacity)
+
+	expVolMaxCapacity := &ap_api.LabelSelectorRequirement{
+		Key:      "px_pool_stats_total_bytes/ 1000000000",
+		Operator: "Lt",
+		Values:   []string{strconv.FormatInt(apParams.PoolScaleMaxSize, 10)},
+	}
+	obj.Spec.Conditions.Expressions = append(obj.Spec.Conditions.Expressions, expVolMaxCapacity)
+
+	return obj, nil
+}
+
+func (d *portworx) createAutopilotRule(autopilotRule *ap_api.AutopilotRule) error {
+	err := d.schedOps.CreateAutopilotRule(autopilotRule)
+	if errors.IsAlreadyExists(err) {
+		logrus.Infof("Found existing AutopilotRule: %v", autopilotRule.Name)
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to create AutopilotRule: %s. Err: %v", autopilotRule.Name, err)
+	}
+
+	return nil
+}
+
```

Comment:
move this logic to driver/scheduler/k8s/k8s.go, the same as I did for for VolumeResize. Looks like you didn't apply my changes from AUT-13 as I already have create createAutopilotRule function there.

Reply:
in this case this is not related to scheduler but the volume driver, @harsh-px found better put under volume driver

Reply:
In line 1776 fix typo penstorage, should be openstorage